src/Storages/Kafka/StorageKafka2.cpp

```diff
@@ -0,0 +1,1291 @@
+        removeTableNodesFromZooKeeper(my_keeper, drop_lock);
+    }
+}
+
+std::optional<StorageKafka2::TopicPartitionLocks>
+StorageKafka2::lockTopicPartitions(zkutil::ZooKeeper & keeper_to_use, const TopicPartitions & topic_partitions)
+{
+    std::vector<fs::path> topic_partition_paths;
+    topic_partition_paths.reserve(topic_partitions.size());
+    for (const auto & topic_partition : topic_partitions)
+        topic_partition_paths.emplace_back(getTopicPartitionPath(topic_partition));
+
+    Coordination::Requests ops;
+
+    static constexpr auto ignore_if_exists = true;
+
+    for (const auto & topic_partition_path : topic_partition_paths)
+    {
+        const auto lock_file_path = String(topic_partition_path / lock_file_name);
+        LOG_TRACE(log, "Creating locking ops for: {}", lock_file_path);
+        ops.push_back(zkutil::makeCreateRequest(topic_partition_path, "", zkutil::CreateMode::Persistent, ignore_if_exists));
+        ops.push_back(zkutil::makeCreateRequest(lock_file_path, kafka_settings->kafka_replica_name.value, zkutil::CreateMode::Ephemeral));
+    }
+    Coordination::Responses responses;
+
+    if (const auto code = keeper_to_use.tryMulti(ops, responses); code != Coordination::Error::ZOK)
+    {
+        if (code != Coordination::Error::ZNODEEXISTS)
+            zkutil::KeeperMultiException::check(code, ops, responses);
+
+        // Possible optimization: check the content of lock files, if we locked them, then we can clean them up and retry to lock them.
+        return std::nullopt;
+    }
+
+    // We have the locks, let's gather the information we needed
+    TopicPartitionLocks locks;
+    {
+        auto tp_it = topic_partitions.begin();
+        auto path_it = topic_partition_paths.begin();
+        for (; tp_it != topic_partitions.end(); ++tp_it, ++path_it)
+        {
+            using zkutil::EphemeralNodeHolder;
+            LockedTopicPartitionInfo lock_info{
+                EphemeralNodeHolder::existing(*path_it / lock_file_name, keeper_to_use),
+                getNumber(keeper_to_use, *path_it / commit_file_name),
+                getNumber(keeper_to_use, *path_it / intent_file_name)};
+
+            LOG_TRACE(
+                log,
+                "Locked topic partition: {}:{} at offset {} with intent size {}",
+                tp_it->topic,
+                tp_it->partition_id,
+                lock_info.committed_offset.value_or(0),
+                lock_info.intent_size.value_or(0));
+            locks.emplace(TopicPartition(*tp_it), std::move(lock_info));
+        }
+    }
+
+    return locks;
+}
+
+
+void StorageKafka2::saveCommittedOffset(zkutil::ZooKeeper & keeper_to_use, const TopicPartition & topic_partition)
+{
+    const auto partition_prefix = getTopicPartitionPath(topic_partition);
+    keeper_to_use.createOrUpdate(partition_prefix / commit_file_name, toString(topic_partition.offset), zkutil::CreateMode::Persistent);
```

Comment:
I'd make update of `partition_prefix / commit_file_name` in a single keeper transaction with checking that partition lock ephemeral node still exists. Also I'd suggest to make a separate persistent node, which will be used as a version tracker of ephemeral partition lock node, e.g. its zversion will be used to verify that ephemeral partition lock was created by the thread which now tries to do the commit.

Reply:
After another thought - this is redundant, as you manage keeper client yourself in this class.

Reply:
Just to clarify: I understood your last comment as current way of handling this is fine, and I don't have to change anything. If I misunderstood you, please let me know.